File: backend/app/models/student.py
"""
Student document model.

Embeddings are NOT stored here — they live in the face_embeddings collection.
This keeps the student document lightweight for listing and search operations.
"""

from typing import Optional
from pydantic import BaseModel, Field
from app.models.base import DocumentBase


class ContactInfo(BaseModel):
    """Nested contact information."""
    phone: str = ""
    email: str = ""


class FaceRegistration(BaseModel):
    """Face registration metadata (NOT the actual embedding)."""
    image_filenames: list[str] = Field(default_factory=list)
    registration_status: str = "pending_image"  # pending_image | processing | active | failed
    registered_at: Optional[str] = None
    image_count: int = 0


class StudentDocument(DocumentBase):
    """MongoDB document model for the students collection."""
    roll_no: str
    name: str
    department: str = "CSE"
    section: str = "A"
    year: str = ""
    semester: Optional[int] = None

    contact_info: ContactInfo = Field(default_factory=ContactInfo)
    face: FaceRegistration = Field(default_factory=FaceRegistration)

    # Denormalized counters (updated atomically via $inc)
    violations_count: int = 0
    late_count: int = 0
    bunk_count: int = 0
    dress_code_count: int = 0
    attendance_percentage: float = 0.0

    # Flexible metadata for future fields
    metadata: dict = Field(default_factory=dict)

File: frontend/src/types/student.ts
export interface ContactInfo {
  phone: string;
  email: string;
}

export interface FaceRegistration {
  image_filenames: string[];
  registration_status: 'pending_image' | 'processing' | 'active' | 'failed';
  image_count: number;
}

export interface Student {
  _id?: string;
  roll_no: string;
  name: string;
  department: string;
  section: string;
  year?: string;
  academic_year?: string;
  current_semester?: string;
  semester_violations?: Record<string, number>;
  contact_info?: ContactInfo;
  face?: FaceRegistration;
  violations_count: number;
  late_count: number;
  bunk_count: number;
  dress_code_count: number;
  attendance_percentage: number;
  status: string;
  created_at?: string;
  updated_at?: string;
}

export interface StudentAnalytics {
  total: number;
  monthly_counts: {
    labels: string[];
    data: number[];
  };
  breakdown: Record<string, number>;
  timeline: Array<{
    id: string;
    type: string;
    date: string;
    remark: string;
    location: string;
    status: string;
  }>;
}
